crates/sui-types/src/messages.rs

```diff
@@ -1052,7 +1095,38 @@ impl TransactionData {
     }
 
     pub fn validity_check(&self) -> SuiResult {
-        Self::validity_check_impl(&self.kind, &self.gas_payment)
+        Self::validity_check_impl(&self.kind, self.gas_payment_object_ref())?;
+        self.check_sponsorship()
+    }
+
+    /// Check if the transaction is compliant with sponsorship.
+    fn check_sponsorship(&self) -> SuiResult {
+        // Not a sponsored transaction, nothing to check
+        if self.gas_owner() == self.sender() {
+            return Ok(());
+        }
+        let allow_sponsored_tx = match &self.kind {
+            // Note: the check for Batch is based on the assumption that
+            // Batch transaction has the same or more restricted
+            // transaction kind limits.
+            TransactionKind::Batch(_b) => true,
```

Comment:
I see that the verification for batch is stricter than this, but that is contingent on how the code is today. If we later expand what is allowed in batch txns, it will be easy to forget to update this check.

I would suggest simply disallowing sponsored batch txns (which have been a source of input validation bugs in the past), unless there is a strong use case for them, in which case lets recursively verify the sub-txns according to the same rules here.

Reply:
fwiw @tnowacki is removing batch txns with programable txns iirc

Reply:
We can start with disallowing sponsored batch tx. I'm not sure it's even used widely today. Will change

Reply:
now we disallow sponsored batch tx